control-plane/bfruntime.h

```diff
@@ -0,0 +1,497 @@
+#ifndef CONTROL_PLANE_BFRUNTIME_H_
+#define CONTROL_PLANE_BFRUNTIME_H_
+
+#include <iosfwd>
+#include <algorithm>
+#include <iomanip>
+#include <iterator>
+#include <limits>
+#include <ostream>
+#include <regex>
+#include <sstream>
+#include <string>
+
+#include <boost/optional.hpp>
+
+#include "control-plane/p4RuntimeSerializer.h"
+#include "lib/exceptions.h"
+#include "lib/gmputil.h"
+#include "lib/json.h"
+#include "lib/log.h"
+#include "lib/null.h"
+#include "p4/config/v1/p4info.pb.h"
+
+namespace p4configv1 = ::p4::config::v1;
+
+namespace P4 {
+struct P4RuntimeAPI;
+}  // namespace P4
+
+namespace P4 {
+
+namespace BFRT {
+
+using P4Id = uint32_t;
+
+// Helpers
+template <typename T, typename R>
+static constexpr P4Id makeBFRuntimeId(T base, R prefix) {
+    return static_cast<P4Id>((base & 0xffffff) | (prefix << 24));
+}
+
+static constexpr P4Id getIdPrefix(P4Id id) {
+    return ((id >> 24) & 0xff);
+}
+
+static Util::JsonObject* findJsonTable(Util::JsonArray* tablesJson, cstring tblName) {
+    for (auto *t : *tablesJson) {
+        auto *tblObj = t->to<Util::JsonObject>();
+        auto tName = tblObj->get("name")->to<Util::JsonValue>()->getString();
+        if (tName == tblName) {
+            return tblObj;
+        }
+    }
+    return nullptr;
+}
+
+// See https://stackoverflow.com/a/33799784/4538702
```

Comment:
perhaps this should be in a library, such as json.h?

Reply:
BTW: there is already an escapeJson in cstring, but it's not exactly the same. Should that one be replaced?